File: src/frontend/src/types/schema-import.ts
export interface BrowseNode {
  name: string;
  node_type: string;
  path: string;
  has_children: boolean;
  description: string | null;
  asset_type: string | null;
  connector_type: string | null;
}

export interface BrowseResponse {
  connection_id: string;
  path: string | null;
  nodes: BrowseNode[];
}

export type ImportDepth = 'selected_only' | 'one_level' | 'full_recursive';

export interface ImportRequest {
  connection_id: string;
  selected_paths: string[];
  depth: ImportDepth;
  dry_run?: boolean;
}

export interface ImportPreviewItem {
  path: string;
  name: string;
  asset_type: string;
  will_create: boolean;
  existing_asset_id: string | null;
  parent_path: string | null;
}

export interface ImportResultItem {
  path: string;
  name: string;
  asset_type: string;
  action: 'created' | 'skipped' | 'error';
  asset_id: string | null;
  error: string | null;
  parent_path: string | null;
}

export interface ImportResult {
  created: number;
  skipped: number;
  errors: number;
  error_messages: string[];
  items: ImportResultItem[];
}

File: src/backend/src/models/schema_import.py
"""
Schema Import Models

Request/response models for the Schema Importer feature, which bridges
external connectors (BigQuery, Databricks, Snowflake, etc.) with persisted
Ontos assets.
"""

from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Browse models
# ---------------------------------------------------------------------------

class BrowseNode(BaseModel):
    """A single node in the remote system hierarchy."""
    name: str = Field(..., description="Display name of the node")
    node_type: str = Field(..., description="Type label (e.g. catalog, schema, dataset, table, view)")
    path: str = Field(..., description="Dot-separated path used for further browsing")
    has_children: bool = Field(False, description="Whether expanding this node will yield children")
    description: Optional[str] = Field(None, description="Optional description/comment")
    asset_type: Optional[str] = Field(None, description="UnifiedAssetType value if this is a leaf asset")
    connector_type: Optional[str] = Field(None, description="Connector type that owns this node")


class BrowseResponse(BaseModel):
    """Response for a browse request."""
    connection_id: UUID
    path: Optional[str] = None
    nodes: List[BrowseNode] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Import depth
# ---------------------------------------------------------------------------

class ImportDepth(str, Enum):
    SELECTED_ONLY = "selected_only"
    ONE_LEVEL = "one_level"
    FULL_RECURSIVE = "full_recursive"


# ---------------------------------------------------------------------------
# Import request / preview / result
# ---------------------------------------------------------------------------

class ImportRequest(BaseModel):
    """Request to preview or execute a schema import."""
    connection_id: UUID = Field(..., description="ID of the connection to import from")
    selected_paths: List[str] = Field(..., min_length=1, description="Dot-separated paths of selected nodes")
    depth: ImportDepth = Field(ImportDepth.FULL_RECURSIVE, description="How deep to recurse below selected nodes")
    dry_run: bool = Field(False, description="If true, only preview — do not create assets")


class ImportPreviewItem(BaseModel):
    """A single item in an import preview."""
    path: str = Field(..., description="Full path of the item")
    name: str = Field(..., description="Display name")
    asset_type: str = Field(..., description="Mapped Ontos asset type name (e.g. Table, View, Column)")
    will_create: bool = Field(True, description="True if a new asset will be created, False if it already exists")
    existing_asset_id: Optional[UUID] = Field(None, description="ID of the existing asset if will_create is False")
    parent_path: Optional[str] = Field(None, description="Path of the parent node (for relationship creation)")


class ImportResultItem(BaseModel):
    """Outcome for a single imported item."""
    path: str
    name: str
    asset_type: str
    action: str = Field(..., description="created | skipped | error")
    asset_id: Optional[UUID] = None
    error: Optional[str] = None
    parent_path: Optional[str] = Field(None, description="Path of the parent node (for tree display)")


class ImportResult(BaseModel):
    """Overall import result."""
    created: int = 0
    skipped: int = 0
    errors: int = 0
    error_messages: List[str] = Field(default_factory=list)
    items: List[ImportResultItem] = Field(default_factory=list)
